test/Rx/Functional/Operator/RetryWhenTest.php

```diff
@@ -0,0 +1,271 @@
+            onError(260, $error)
+        ], $results->getMessages());
+
+        $this->assertSubscriptions([
+            subscribe(200, 230),
+            subscribe(230, 260)
+        ], $xs->getSubscriptions());
+    }
+
+    /**
+     * @test
+     */
+    public function retryWhen_Observable_complete()
+    {
+        $error = new \Exception();
+
+        $xs = $this->createColdObservable([
+            onNext(10, 1),
+            onNext(20, 2),
+            onError(30, $error),
+            onCompleted(40)
+        ]);
+
+        $results = $this->scheduler->startWithCreate(function () use ($xs, $error) {
+            return $xs->retryWhen(function (Observable $attempts) {
+                return Observable::emptyObservable();
+            });
+        });
+
+        $this->assertMessages([
+            onNext(210, 1),
+            onNext(220, 2),
+            onCompleted(230)
+        ], $results->getMessages());
+
+        $this->assertSubscriptions([
+            subscribe(200, 230)
+        ], $xs->getSubscriptions());
+    }
+
+    /**
+     * @test
+     */
+    public function retryWhen_Observable_next_complete()
+    {
+        $this->markTestSkipped("skip until takeWhile is implemented");
+
+        $error = new \Exception();
+
+        $xs = $this->createColdObservable([
+            onNext(10, 1),
+            onNext(20, 2),
+            onError(30, $error),
+            onCompleted(40)
+        ]);
+
+        $results = $this->scheduler->startWithCreate(function () use ($xs, $error) {
+            return $xs->retryWhen(function (Observable $attempts) {
+                return $attempts->scan(function ($count, $error) {
+                    return $count + 1;
+                }, 0)->takeWhile(function ($count) {
+                    return $count < 2;
+                });
+            });
+        });
+
+        $this->assertMessages([
+            onNext(210, 1),
+            onNext(220, 2),
+            onNext(240, 1),
+            onNext(250, 2),
+            onCompleted(260)
+        ], $results->getMessages());
+
+        $this->assertSubscriptions([
+            subscribe(200, 230),
+            subscribe(230, 260)
+        ], $xs->getSubscriptions());
+    }
+
+    /**
+     * @test
+     */
+    public function retryWhen_Observable_infinite()
+    {
+        $error = new \Exception();
+
+        $xs = $this->createColdObservable([
+            onNext(10, 1),
+            onNext(20, 2),
+            onError(30, $error),
+            onCompleted(40)
+        ]);
+
+        $results = $this->scheduler->startWithCreate(function () use ($xs, $error) {
+            return $xs->retryWhen(function () {
+                return Observable::never();
+            });
+        });
+
+        $this->assertMessages([
+            onNext(210, 1),
+            onNext(220, 2)
+        ], $results->getMessages());
+
+        $this->assertSubscriptions([
+            subscribe(200, 230)
+        ], $xs->getSubscriptions());
+    }
```

Comment:
I used the tests from the RxJS - I think there could be a few more cases added. I will work on that when I have a chance.